frame/system/src/lib.rs

```diff
@@ -847,6 +847,27 @@ impl<O, T> EnsureOrigin<O> for EnsureNever<T> {
 	}
 }
 
+pub struct EnsureOneOf<AccountId, A, B>(A, B, sp_std::marker::PhantomData<AccountId>);
+impl<
+	AccountId,
+	O: Into<Result<RawOrigin<AccountId>, O>> + From<RawOrigin<AccountId>>,
+	A: EnsureOrigin<O>,
+	B: EnsureOrigin<O>,
+> EnsureOrigin<O> for EnsureOneOf<AccountId, A, B> {
+	type Success = ();
```

Comment:
I would change this `Success` to:

```
enum Either<L, R> {
    Left(L),
    Right(R),
}

type Success = Either<A::Success, B::Success>;
```

Reply:
That or require both to have the same `Success` type and return that. Can be constrained using `B: EnsureOrigin<O, Success=A::Success>`.

The thing I worry about with an either type is that there are loads of different definitions of `Either`. I'd recommend re-exporting the type from https://crates.io/crates/either in `no_std` mode from either `frame-support` or `sp-runtime` if an `Either` is used rather than redefining it.

Reply:
I think we should not require the same `Success type.

Regarding using the `either` crate, I thought the same. However I was not entirely sure. In the end it is a simple enum. But I also don't care :shrug: